docs/training/apis-howto/deepspeed/autotuning.rst

```diff
@@ -0,0 +1,247 @@
+.. _deepspeed-autotuning:
+
+#######################################
+DeepSpeed Autotune
+#######################################
+
+.. meta::
+   :description: This user guide demonstrates how to optimize DeepSpeed parameter in order to take full advantage of the user's hardware and model.
+
+
+Getting the most out of DeepSpeed (DS) requires aligning the many DS parameters with the specific
+properties of your hardware and model. Determined AI's DeepSpeed Autotune (``dsat``) helps to optimize these
+settings through an easy-to-use API with very few changes required in user-code, as we describe
+in the remainder of this user guide. ``dsat`` can be used with any of
+:class:`~determined.pytorch.deepspeed.DeepSpeedTrial`, Core API, and HuggingFace Trainer.
+ 
+
+
+What to Expect
+================
+
+Assuming you have DeepSpeed code which already functions, autotuning is as easy as inserting one or
+two helper functions into your code and then replacing the usual experiment launching command 
+
+.. code::
+
+  det experiment create deepspeed.yaml .
+
+with
+
+.. code::
+
+  python3 -m determined.pytorch.dsat asha deepspeed.yaml .
+
+where ``deepspeed.yaml`` is the Determined configuration for a ``single`` experiment. There is no
+need to write a special configuration file to use ``dsat``.
+
+The above uses the ASHA algorithm (TODO: link to ASHA page) to tune the DS parameters and is one
+of the three search methods we currently provide:
+
+* ``asha``: adaptively searches over randomly selected DeepSpeed configurations, yielding more
+   compute resources to well-performing classes of configurations.
+
+* ``binary``: simple binary searches over the batch size for randomly-generated DS configurations.
+
+* ``random``: a search over random DeepSpeed configurations with an aggressive early-stopping
+  criteria based on domain-knowledge of DeepSpeed and the search history.
+
+DeepSpeed Autotune is built on top of Custom Searcher (TODO: link) 
+which starts up two separate experiments:
+
+* A ``single`` search runner experiment whose role is to coordinate and schedule the actual trials
+  which run model code. 
+* A ``custom`` experiment which contains the trials referenced above whose results are reported back
+  to the search runner.
+
+The logs of the ``single`` search runner experiment will contain information regarding the size of the model,
+the GPU memory available, the activation memory required per example, and an approximate computation
+of the maximum batch size per zero stage (which is used to guide the starting point for all
+autotuning searches). When a best-performing DS configuration is found, the
+corresponding ``json`` configuration file will be written to the search runner's checkpoint directory,
+along with a file detailing the configuration's corresponding metrics.
```

Comment:
Do we need to tell users this much detail about the guts of how the autotuning works? Maybe this could be explained in a section dedicated to a deep dive? It feels like too much detail too soon to me